packages/http-server-javascript/helpers/router.ts

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation
+// Licensed under the MIT license.
+
+import type * as http from "http";
+
+/** A policy that can be applied to a route or a set of routes. */
+export interface Policy {
+  /** Optional policy name. */
+  name?: string;
+
+  /**
+   * Applies the policy to the request.
+   *
+   * Policies _MUST_ call `next()` to pass the request to the next policy _OR_ call `response.end()` to terminate,
+   * and _MUST NOT_ do both.
+   *
+   * If the policy passes a `request` object to `next()`, that request object will be used instead of the original
+   * request object for the remainder of the policy chain. If the policy does _not_ pass a request object to `next()`,
+   * the same object that was passed to this policy will be forwarded to the next policy automatically.
+   *
+   * @param request - The incoming HTTP request.
+   * @param response - The outgoing HTTP response.
+   * @param next - Calls the next policy in the chain.
+   */
+  (
+    request: http.IncomingMessage,
+    response: http.ServerResponse,
+    next: (request?: http.IncomingMessage) => void
+  ): void;
+}
+
+/**
+ * Create a function from a chain of policies.
+ *
+ * This returns a single function that will apply the policy chain and eventually call the provided `next()` function.
+ *
+ * @param name - The name to give to the policy chain function.
+ * @param policies - The policies to apply to the request.
+ * @param out - The function to call after the policies have been applied.
+ */
+export function createPolicyChain<
+  Out extends (
+    ctx: HttpContext,
+    request: http.IncomingMessage,
+    response: http.ServerResponse,
+    ...rest: any[]
+  ) => void,
+>(name: string, policies: Policy[], out: Out): Out {
+  let outParams: any[];
+  if (policies.length === 0) {
+    return out;
+  }
+
+  function applyPolicy(
+    ctx: HttpContext,
+    request: http.IncomingMessage,
+    response: http.ServerResponse,
+    index: number
+  ) {
+    if (index >= policies.length) {
+      return out(ctx, request, response, ...outParams);
+    }
+
+    policies[index](request, response, function nextPolicy(nextRequest) {
+      applyPolicy(ctx, nextRequest ?? request, response, index + 1);
+    });
+  }
+
+  return {
+    [name](
+      ctx: HttpContext,
+      request: http.IncomingMessage,
+      response: http.ServerResponse,
+      ...params: any[]
+    ) {
+      outParams = params;
+      applyPolicy(ctx, request, response, 0);
+    },
+  }[name] as Out;
+}
+
+/**
+ * The type of an error encountered during request validation.
+ *
+ * TODO/witemple: Need to define shape of error object in a common location.
```

Comment:
todo?

Reply:
There are some TODOs around the code. I will create issues for them once we merge, or I could do it before merging. Either way. This particular one I guess is more important to resolve before merging since it will end up part of the generated output, so I'll create a tracking issue and remove this particular instance anyway.